config/dependency.go

```diff
@@ -418,6 +429,59 @@ func canGetRemoteState(remoteState *remote.RemoteState) bool {
 	return remoteState != nil && !remoteState.DisableDependencyOptimization
 }
 
+// terragruntAlreadyInit returns true if it detects that the module specified by the given terragrunt configuration is
+// already initialized with the terraform source. This will also return the working directory where you can run
+// terraform.
+func terragruntAlreadyInit(terragruntOptions *options.TerragruntOptions, configPath string) (bool, string, error) {
+	// We need to first determine the working directory where the terraform source should be located. This is dependent
+	// on the source field of the terraform block in the config.
+	terraformBlockTGConfig, err := PartialParseConfigFile(configPath, terragruntOptions, nil, []PartialDecodeSectionType{TerraformSource})
+	if err != nil {
+		return false, "", err
+	}
+	var workingDir string
+	sourceUrl := GetTerraformSourceUrl(terragruntOptions, terraformBlockTGConfig)
+	if sourceUrl == "" || sourceUrl == "." {
+		// When there is no source URL, there is no download process and the working dir is the same as the directory
+		// where the config is.
+		if util.IsDir(configPath) {
+			workingDir = configPath
+		} else {
+			workingDir = filepath.Dir(configPath)
+		}
+	} else {
+		terraformSource, err := tfsource.NewTerraformSource(sourceUrl, terragruntOptions.DownloadDir, terragruntOptions.WorkingDir, terragruntOptions.Logger)
+		if err != nil {
+			return false, "", err
+		}
+		// We're only interested in the computed working dir.
+		workingDir = terraformSource.WorkingDir
+	}
+	// Terragrunt is already init-ed if the terraform state dir (.terraform) exists in the working dir.
```

Comment:
Does this cause problems if some `ref` parameter changed? Or a `provider` version changed? That is, if the user has a `.terraform` folder in that folder, but still needs to force `init` to run, is there a way to do that?

Reply:
In this case, the working dir would be different, and thus it won't detect that the module is already `init`-ed. This is because the [working dir depends on the download dir which depends on the ref](https://github.com/gruntwork-io/terragrunt/blob/yori-faster-outputs/cli/tfsource/types.go#L118).

Reply:
Added comment clarifying this: 92e9001

Reply:
Right... But what about the intermittent errors where Terraform fails to download a provider? The `.terraform` folder exists, and `ref` did not change, but yet, you need to run `init` again. Is there a way to force that? I ask because, sadly, these intermittent errors happen quite often, esp with `xxx-all` commands.

Reply:
Ah gotcha. I see what you are getting at. Let me do a few experiments to see if we can autodetect this, but I think an easy workaround is for someone to run `terragrunt run-all init`.

Reply:
UPDATE: it looks like you can still retrieve outputs even if you enter a state where you need to reinit. Here's what I tried to see this behavior:

- Created a module with a single resource:
```
provider "null" {
  version = "~> 2.0"
}

resource "null_resource" "foo" {}

output "foo" {
  value = null_resource.foo.id
}
```

- Ran `terraform init && terraform apply`. Verified outputs can be retrieved with `terraform output`
- Updated the `version` field for the `null` provider. Verified outputs can be retrieved with `terraform output`.
- Added a module block, and confirmed `plan` fails (`Error: Module not installed`). However, outputs were retrievable even without init with `terraform output`.

Reply:
Roger, thanks for checking! That seems good enough to me. If some issue does arise, then you're right, `run-all init` is prob good enough. Thx!